File: python/String/RemoveOutermostParanthesis.py
def removeOuterParentheses(S):
    # res=""
    # count=0
    # for c in S:
    #     if (c=="(" and count>0):
    #         res+=c
    #     if(c=="("):
    #         count+=1
    #     if (c==")" and count>1):
    #         res+=c
    #     if(c==")"):
    #         count-=1
    # return res
    
    opencount=0,closecount=0
    res=""
    start=0
    for i,c in enumerate(S):
        if c=="(":
            opencount+=1
        else:
            closecount+=1
        if opencount==closecount:
            res+=S[start+1:i]
            start=i+1
    return res
n=input()
print(removeOuterParentheses(n))

File: cpp/String/RemoveOutermostParanthesis.cpp
#include<bits/stdc++.h>
using namespace std;

string removeOuterParentheses(string s)
{
    // string ans;
    // int balance=0;
    // for(char c:s)
    // {
    //     if(c=='(' && balance++>0) ans+=c;
    //     if(c==')' && balance-->1) ans+=c;
    // }
    // return ans;

    int opencount=0,closecount=0;
    string res="";
    int start=0;
    for(int i=0;i<s.length();i++)
    {
        char c=s[i];
        if(c=='(')
        {
            opencount++;
        }
        else
        {
            closecount++;
        }
        if(opencount==closecount)
        {
            res+=s.substr(start+1,i-start-1);
            start=i+1;
        }
    }
    return res;
}

int main()
{
    string s;
    cin>>s;
    cout<<removeOuterParentheses(s);
}

//TC=O(N) SC=O(N)
//TC=O(N) SC=O(1)